README.md
# jupyterhub-ldap-authenticator

LDAP Authenticator plugin for [JupyterHub](https://github.com/jupyterhub/jupyterhub).
This project was written with Enterprise LDAP integration in mind and includes the
following features:

- Supports multiple LDAP servers and allows for configuration of `server_pool_strategy`
- Uses single read-only LDAP connection per authentication request
- Verifies authenticating user exists in LDAP and is a member of `allowed_groups`
    before testing authentication
- Supports using nested groups in `allowed_groups` list
- Supports domain user home directory creation at login

This project was inspired by the [ldapauthenticator](https://github.com/jupyterhub/ldapauthenticator) project

## Installation

Install with pip:

```
pip install jupyterhub-ldap-authenticator
```

## Configuration

To enable LDAPAuthenticator, add the following line to the Jupyterhub config file and extend configuration with the parameters listed below.

```python
c.JupyterHub.authenticator_class = 'ldapauthenticator.LDAPAuthenticator'
```

<dl>
  <dt>LDAPAuthenticator.server_hosts</dt>
  <dd>List of Names, IPs, or the complete URLs in the scheme://hostname:hostport format of the server (required).</dd>
</dl>

```python
# example- list of complete urls
c.LDAPAuthenticator.server_hosts = ['ldaps://ldap1.example.com:636', 'ldaps://ldap2.example.com:636']

# example - list of names
c.LDAPAuthenticator.server_hosts = ['ldap1.example.com', 'ldap2.example.com']

# example - list of ips
c.LDAPAuthenticator.server_hosts = ['10.0.0.1', '10.0.0.2']
```

<dl>
  <dt>LDAPAuthenticator.server_port</dt>
  <dd>The port where the LDAP server is listening. Typically 389, for a cleartext connection, and 636 for a secured connection (defaults to None).</dd>
</dl>

```python
# example
c.LDAPAuthenticator.server_port = 636
  ```

<dl>
  <dt>LDAPAuthenticator.server_use_ssl</dt>
  <dd>Boolean specifying if the connection is on a secure port (defaults to False).</dd>
</dl>

```python
# example
c.LDAPAuthenticator.server_use_ssl = True
```

<dl>
  <dt>LDAPAuthenticator.server_connect_timeout</dt>
  <dd>Timeout in seconds permitted when establishing an ldap connection before raising an exception (defaults to None).</dd>
</dl>

```python
# example
c.LDAPAuthenticator.server_connect_timeout = 10
```

<dl>
  <dt>LDAPAuthenticator.server_receive_timeout</dt>
  <dd>Timeout in seconds permitted for responses from established ldap connections before raising an exception (defaults to None).</dd>
</dl>

```python
# example
c.LDAPAuthenticator.server_receive_timeout = 10
```

<dl>
  <dt>LDAPAuthenticator.server_pool_strategy</dt>
  <dd>Available Pool HA strategies (defaults to 'FIRST').</dd>
</dl>

  - FIRST: Gets the first server in the pool, if 'server_pool_active' is set to True gets the first available server.
  - ROUND_ROBIN: Each time the connection is open the subsequent server in the pool is used. If 'server_pool_active' is set to True unavailable servers will be discarded.
  - RANDOM: each time the connection is open a random server is chosen in the pool. If 'server_pool_active' is set to True unavailable servers will be discarded.

```python
# example
c.LDAPAuthenticator.server_pool_strategy = 'FIRST'
```

<dl>
  <dt>LDAPAuthenticator.server_pool_active</dt>
  <dd>If True the ServerPool strategy will check for server availability. Set to Integer for maximum number of cycles to try before giving up (defaults to True).</dd>
</dl>

```python
# example - boolean
c.LDAPAuthenticator.server_pool_active = True

# example - maximum number of tries
c.LDAPAuthenticator.server_pool_active = 3
```

<dl>
  <dt>LDAPAuthenticator.server_pool_exhaust</dt>
  <dd>If True, any inactive servers will be removed from the pool. If set to an Integer, this will be the number of seconds an unreachable server is considered offline. When this timeout expires the server is reinserted in the pool and checked again for availability (defaults to False).</dd>
</dl>

```python
# example - boolean
c.LDAPAuthenticator.server_pool_exhaust = True

# example - offline timeout
c.LDAPAuthenticator.server_pool_exhaust = 600
```

<dl>
  <dt>LDAPAuthenticator.bind_user_dn</dt>
  <dd>The account of the user to log in for simple bind (defaults to None).</dd>
</dl>

```python
# example - freeipa
c.LDAPAuthenticator.bind_user_dn = 'uid=imauser,cn=users,cn=accounts,dc=example,dc=com'

# example - Active Directory
c.LDAPAuthenticator.bind_user_dn = 'CN=imauser,CN=Users,DC=example,DC=com'
```

<dl>
  <dt>LDAPAuthenticator.bind_user_password</dt>
  <dd>The password of the user for simple bind (defaults to None).</dd>
</dl>

```python
# example
c.LDAPAuthenticator.bind_user_password = 'password'
```

<dl>
  <dt>LDAPAuthenticator.user_search_base</dt>
  <dd>The location in the Directory Information Tree where the user search will start.</dd>
</dl>

```python
# example - freeipa
c.LDAPAuthenticator.user_search_base = 'cn=users,cn=accounts,dc=example,dc=com'

# example - active directory
c.LDAPAuthenticator.user_search_base = 'CN=Users,DC=example,DC=com'
```

<dl>
  <dt>LDAPAuthenticator.user_search_filter</dt>
  <dd>LDAP search filter to validate that the authenticating user exists within the organization. Search filters containing '{username}' will have that value substituted with the username of the authenticating user.</dd>
</dl>

```python
# example - freeipa
c.LDAPAuthenticator.user_search_filter = '(&(objectClass=person)(uid={username}))'

# example - active directory
c.LDAPAuthenticator.user_search_filter = '(&(objectCategory=person)(objectClass=user)(sAMAccountName={username}))'
```

<dl>
  <dt>LDAPAuthenticator.user_membership_attribute</dt>
  <dd>LDAP Attribute used to associate user group membership (defaults to 'memberOf').</dd>
</dl>

```python
# example
c.LDAPAuthenticator.user_membership_attribute = 'memberOf'
```

<dl>
  <dt>LDAPAuthenticator.group_search_base</dt>
  <dd>The location in the Directory Information Tree where the group search will start.
  Search string containing '{group}' will be substituted with entries taken from
  allowed_groups</dd>
</dl>

```python
# example - freeipa
c.LDAPAuthenticator.group_search_base = 'cn=groups,cn=accounts,dc=example,dc=com'

# example - active directory
c.LDAPAuthenticator.group_search_base = 'CN=Groups,DC=example,DC=com'
```

<dl>
  <dt>LDAPAuthenticator.group_search_filter</dt>
  <dd>LDAP search filter to return members of groups defined in the allowed_groups parameter. Search filters containing '{group}' will have that value substituted with the group dns provided in the allowed_groups parameter.</dd>
</dl>

```python
# example - freeipa
c.LDAPAuthenticator.group_search_filter = '(&(objectClass=ipausergroup)(memberOf={group}))'

# example - active directory
c.LDAPAuthenticator.group_search_filter = '(&(objectClass=group)(memberOf={group}))'
```

<dl>
  <dt>LDAPAuthenticator.allowed_groups</dt>
  <dd>List of LDAP group DNs that users must be a member of in order to be granted login. If left undefined or set to None, allowed_groups will be short-circuited and all users will be allowed (defaults to None).</dd>
</dl>

```python
# example
c.LDAPAuthenticator.allowed_groups = ['cn=jupyterhub-users,cn=groups,cn=accounts,dc=example,dc=com']
```

<dl>
  <dt>LDAPAuthenticator.allow_nested_groups</dt>
  <dd>Boolean allowing for recursive search of members within nested groups of
  allowed_groups (defaults to False).</dd>
</dl>

```python
# example
c.LDAPAuthenticator.allow_nested_groups = True
```

<dl>
  <dt>LDAPAuthenticator.username_pattern</dt>
  <dd>Regular expression pattern that all valid usernames must match. If a username
  does not match the pattern specified here, authentication will not be attempted.
  If not set, allow any username (defaults to None).</dd>
</dl>

```python
# example - freeipa
c.LDAPAuthenticator.username_pattern = '[a-zA-Z0-9_.][a-zA-Z0-9_.-]{0,252}[a-zA-Z0-9_.$-]?'

# example - active directory
c.LDAPAuthenticator.username_pattern = '[a-zA-Z0-9_.][a-zA-Z0-9_.-]{8,20}[a-zA-Z0-9_.$-]?'
```

<dl>
  <dt>LDAPAuthenticator.create_user_home_dir</dt>
  <dd>Boolean allowing for user home directory to be created at login</dd>
</dl>

```python
# example
c.LDAPAuthenticator.create_user_home_dir = True
```

<dl>
  <dt>LDAPAuthenticator.create_user_home_dir_cmd</dt>
  <dd>Command used when creating a userhome directory as a list of strings. The
  username will be appended as the final argument. Defaults
  to `mkhomedir_helper` on linux systems.</dd>
</dl>

```python
# example
c.LDAPAuthenticator.create_user_home_dir_cmd = ['mkhomedir_helper']
```


## Examples

##### FreeIPA Integration

```python
# freeipa example
c.JupyterHub.authenticator_class = 'ldapauthenticator.LDAPAuthenticator'
c.LDAPAuthenticator.server_hosts = ['ldaps://ldap1.example.com:636', 'ldaps://ldap2.example.com:636']
c.LDAPAuthenticator.bind_user_dn = 'uid=imauser,cn=users,cn=accounts,dc=example,dc=com'
c.LDAPAuthenticator.bind_user_password = 'imapassword'
c.LDAPAuthenticator.user_search_base = 'cn=users,cn=accounts,dc=example,dc=com'
c.LDAPAuthenticator.user_search_filter = '(&(objectClass=person)(uid={username}))'
c.LDAPAuthenticator.user_membership_attribute = 'memberOf'
c.LDAPAuthenticator.group_search_base = 'cn=groups,cn=accounts,dc=example,dc=com'
c.LDAPAuthenticator.group_search_filter = '(&(objectClass=ipausergroup)(memberOf={group}))'
c.LDAPAuthenticator.allowed_groups = ['cn=jupyterhub-users,cn=groups,cn=accounts,dc=example,dc=com']
c.LDAPAuthenticator.allow_nested_groups = True
c.LDAPAuthenticator.username_pattern = '[a-zA-Z0-9_.][a-zA-Z0-9_.-]{0,252}[a-zA-Z0-9_.$-]?'
c.LDAPAuthenticator.create_user_home_dir = True
c.LDAPAuthenticator.create_user_home_dir_cmd = ['mkhomedir_helper']
```

##### Active Directory Integration

```python
# active directory example
c.JupyterHub.authenticator_class = 'ldapauthenticator.LDAPAuthenticator'
c.LDAPAuthenticator.server_hosts = ['ldaps://ldap1.example.com:636', 'ldaps://ldap2.example.com:636']
c.LDAPAuthenticator.bind_user_dn = 'CN=imauser,CN=Users,DC=example,DC=com'
c.LDAPAuthenticator.bind_user_password = 'imapassword'
c.LDAPAuthenticator.user_search_base = 'CN=Users,DC=example,DC=com'
c.LDAPAuthenticator.user_search_filter = '(&(objectCategory=person)(objectClass=user)(sAMAccountName={username}))
c.LDAPAuthenticator.user_membership_attribute = 'memberOf'
c.LDAPAuthenticator.group_search_base = 'CN=Groups,DC=example,DC=com'

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased](unreleased)

- no new features in development at this time

## [0.4.1](https://github.com/hansohn/jupyterhub-ldap-authenticator/compare/0.4.0...0.4.1) (Dev 21, 2020)

BUG FIXES:

- set auto_bind to AUTO_BIND_NO_TLS to account to ldap3 bind changes

## [0.4.0](https://github.com/hansohn/jupyterhub-ldap-authenticator/compare/0.3.3...0.4.0) (Jan 13, 2020)

FEATURES:

- bump python requirement to ~=3.5
- use AUTO_BIND_NO_TLS for non-ldaps connections
- migrate logs to str.format()
- update hostname regex
- remove username normalization
- remove dn attribute filtering on auth_user_search_filter response
- PEP8 updates

## [0.3.3](https://github.com/hansohn/jupyterhub-ldap-authenticator/compare/0.3.2...0.3.3) (Jan 08, 2020)

BUG FIXES:

- filter out user search results not containing the dn attribute
- only verify local user exists if create_user_home_dir is True

## [0.3.2](https://github.com/hansohn/jupyterhub-ldap-authenticator/compare/0.3.1...0.3.2) (Mar 20, 2019)

BUG FIXES:

- update group_search_filter documentation in README
- modify test_auth function to work with OTP passwords
- update rebind error statement

## [0.3.1](https://github.com/hansohn/jupyterhub-ldap-authenticator/compare/0.3.0...0.3.1) (Feb 5, 2019)

BUG FIXES:

- update documentation to reflect correct user_search_filter for active directory integration
- allow server_hosts to use non-standard ldap ports

## [0.3.0](https://github.com/hansohn/jupyterhub-ldap-authenticator/compare/0.2.0...0.3.0) (Nov 1, 2018)

FEATURES:

- set 'allowed_groups' to None to short-circuit user group membership check. For use in LDAP solutions that do not allow user group memberships to be easily queried.

BUG FIXES:

- add try/catch to pwd.getpwnam call
- fix Readme parameter typos
- fix Changelog linking

## [0.2.0](https://github.com/hansohn/jupyterhub-ldap-authenticator/compare/0.1.0...0.2.0) (May 31, 2018)

FEATURES:

- rename keyword substituion keys `user_logon` and `group_dn` to `username` and `group` to align with other ldap authenticator projects for easier transition.

BUG FIXES:

- fix nesting method to return proper results
- fix bug that permitted user login after being removed from allowed_groups memberships
- update code to adheare to pylint standards

## [0.1.0](https://github.com/hansohn/jupyterhub-ldap-authenticator/compare/0.1.0...0.1.0) (May 04, 2018)

FEATURES:

- initial commit


ldapauthenticator/ldapauthenticator.py
# -*- coding: utf-8 -*-
# MIT License
#
# Copyright (c) 2018 Ryan Hansohn
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

"""
LDAP Authenticator plugin for JupyterHub
"""

import os
import pwd
import re
import subprocess
import sys
import typing
from jupyterhub.auth import Authenticator
from jupyterhub.orm import User
from jupyterhub.traitlets import Command
import ldap3
import ldap3.core.exceptions
from tornado import gen, web
from traitlets import Any, Int, Bool, List, Unicode, Union, default, observe


class LDAPAuthenticator(Authenticator):
    """
    LDAP Authenticator for Jupyterhub
    """

    server_hosts = Union(
        [List(), Unicode()],
        config=True,
        help="""
        List of Names, IPs, or the complete URLs in the scheme://host:port
        format of the server (required).
        """
    )

    server_port = Int(
        allow_none=True,
        default_value=None,
        config=True,
        help="""
        The port where the LDAP server is listening. Typically 389, for a
        cleartext connection, and 636 for a secured connection (defaults to None).
        """
    )

    server_use_ssl = Bool(
        default_value=False,
        config=True,
        help="""
        Boolean specifying if the connection is on a secure port (defaults to False).
        """
    )

    server_connect_timeout = Int(
        allow_none=True,
        default_value=None,
        config=True,
        help="""
        Timeout in seconds permitted when establishing an ldap connection before
        raising an exception (defaults to None).
        """
    )

    server_receive_timeout = Int(
        allow_none=True,
        default_value=None,
        config=True,
        help="""
        Timeout in seconds permitted for responses from established ldap
        connections before raising an exception (defaults to None).
        """
    )

    server_pool_strategy = Unicode(
        default_value='FIRST',
        config=True,
        help="""
        Available Pool HA strategies (defaults to 'FIRST').

        FIRST: Gets the first server in the pool, if 'server_pool_active' is
            set to True gets the first available server.
        ROUND_ROBIN: Each time the connection is open the subsequent server in
            the pool is used. If 'server_pool_active' is set to True unavailable
            servers will be discarded.
        RANDOM: each time the connection is open a random server is chosen in the
            pool. If 'server_pool_active' is set to True unavailable servers
            will be discarded.
        """
    )

    server_pool_active = Union(
        [Bool(), Int()],
        default_value=True,
        config=True,
        help="""
        If True the ServerPool strategy will check for server availability. Set
        to Integer for maximum number of cycles to try before giving up
        (defaults to True).
        """
    )

    server_pool_exhaust = Union(
        [Bool(), Int()],
        default_value=False,
        config=True,
        help="""
        If True, any inactive servers will be removed from the pool. If set to
        an Integer, this will be the number of seconds an unreachable server is
        considered offline. When this timeout expires the server is reinserted
        in the pool and checked again for availability (defaults to False).
        """
    )

    bind_user_dn = Unicode(
        allow_none=True,
        default_value=None,
        config=True,
        help="""
        The account of the user to log in for simple bind (defaults to None).
        """
    )

    bind_user_password = Unicode(
        allow_none=True,
        default_value=None,
        config=True,
        help="""
        The password of the user for simple bind (defaults to None)
        """
    )

    user_search_base = Unicode(
        config=True,
        help="""
        The location in the Directory Information Tree where the user search
        will start.
        """
    )

    user_search_filter = Unicode(
        config=True,
        help="""
        LDAP search filter to validate that the authenticating user exists
        within the organization. Search filters containing '{username}' will
        have that value substituted with the username of the authenticating user.
        """
    )

    user_membership_attribute = Unicode(
        default_value='memberOf',
        config=True,
        help="""
        LDAP Attribute used to associate user group membership
        (defaults to 'memberOf').
        """
    )

    group_search_base = Unicode(
        config=True,
        help="""
        The location in the Directory Information Tree where the group search
        will start. Search string containing '{group}' will be substituted
        with entries taken from allow_nested_groups.
        """
    )

    group_search_filter = Unicode(
        config=True,
        help="""
        LDAP search filter to return members of groups defined in the
        allowed_groups parameter. Search filters containing '{group}' will
        have that value substituted with the group dns provided in the
        allowed_groups parameter.
        """
    )

    allowed_groups = Union(
        [Unicode(), List()],
        allow_none=True,
        default_value=None,
        config=True,
        help="""
        List of LDAP group DNs that users must be a member of in order to be granted
        login.
        """
    )

    allow_nested_groups = Bool(
        default_value=False,
        config=True,
        help="""
        Boolean allowing for recursive search of members within nested groups of
        allowed_groups (defaults to False).
        """
    )

    username_pattern = Unicode(
        config=True,
        help="""
        Regular expression pattern that a valid username must match. If a
        username does not match the pattern specified here, authentication will
        not be attempted. If not set, allow any username (defaults to None).
        """
    )

    username_regex = Any(
        help="""
        Compiled regex kept in sync with `username_pattern`
        """
    )

    @observe('username_pattern')
    def _username_pattern_changed(self, change: dict) -> None:
        if not change['new']:
            self.username_regex = None
        self.username_regex = re.compile(change['new'])

    create_user_home_dir = Bool(
        default_value=False,
        config=True,
        help="""
        If set to True, will attempt to create a user's home directory
        locally if that directory does not exist already.
        """
    )

    create_user_home_dir_cmd = Command(
        config=True,
        help="""
        Command to create a users home directory.
        The command should be formatted as a list of strings.
        """
    )

    @default('create_user_home_dir_cmd')
    def _default_create_user_home_dir_cmd(self) -> typing.List[str]:
        if sys.platform == 'linux':
            home_dir_cmd = ['mkhomedir_helper']
        else:
            self.log.debug("Not sure how to create a home directory on '{}' system".format(
                sys.platform))
            home_dir_cmd = list()
        return home_dir_cmd

    @gen.coroutine
    def add_user(self, user: User) -> typing.Generator:
        if self.create_user_home_dir:
            username = user.name
            user_exists = yield gen.maybe_future(self.user_home_dir_exists(username))
            if not user_exists:
                yield gen.maybe_future(self.add_user_home_dir(username))
        yield gen.maybe_future(super().add_user(user))

    def user_home_dir_exists(self, username: str) -> bool:
        """
        Verify user home directory exists
        """
        try:
            user = pwd.getpwnam(username)
            home_dir = user.pw_dir
            return os.path.isdir(home_dir)
        except KeyError:
            return False

    def add_user_home_dir(self, username: str) -> None:
        """
        Creates user home directory
        """
        cmd = self.create_user_home_dir_cmd + [username]
        self.log.info("Creating '{}' user home directory using command '{}'".format(
            username, ' '.join(cmd)))
        proc = subprocess.Popen(cmd,
                                stdout=subprocess.PIPE,
                                stderr=subprocess.PIPE,
                                universal_newlines=True)
        out, err = proc.communicate()
        if proc.returncode:
            raise RuntimeError("Failed to create '{}' user home directory: {}".format(
                username, err))

    def validate_username(self, username: str) -> bool:
        """
        Validate a username
        Return True if username is valid, False otherwise.
        """
        if '/' in username:
            # / is not allowed in usernames
            return False
        if not username:
            # empty usernames are not allowed
            return False
        if not self.username_regex:
            return True
        return bool(self.username_regex.match(username))

    def validate_host(self, host: str) -> bool:
        """
        Validate hostname
        Return True if host is valid, False otherwise.
        """
        ip_address_regex = re.compile(r'^(?:(?:25[0-5]|2[0-4][0-9]|1[0-9][0-9]|[1-9]?[0-9])\.){3}'
                                      r'(?:25[0-5]|2[0-4][0-9]|1[0-9][0-9]|[1-9]?[0-9])$')
        hostname_regex = re.compile(r'^(?:[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?\.)+'
                                    r'[a-z0-9][a-z0-9-]{0,61}[a-z0-9]$')
        url_regex = re.compile(r'^(ldaps?)://'
                               r'((?:[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?\.)+'
                               r'[a-z0-9][a-z0-9-]{0,61}[a-z0-9]):'
                               r'([0-9]{1,5})$')
        if bool(ip_address_regex.match(host)):
            # using ipv4 address
            valid = True
        elif bool(hostname_regex.match(host)):
            # using a hostname address
            valid = True
        elif bool(url_regex.match(host)):
            # using host url address
            match = url_regex.match(host)
            proto = match.group(1)
            if proto == 'ldaps':
                self.server_use_ssl = True
            valid = True
        else:
            # unsupported host format
            valid = False
        return valid

    def create_ldap_server_pool_obj(self,
                                    ldap_servers: typing.List[str] = None) -> ldap3.ServerPool:
        """
        Create ldap3 ServerPool Object
        """
        server_pool = ldap3.ServerPool(
            ldap_servers,
            pool_strategy=self.server_pool_strategy.upper(),
            active=self.server_pool_active,
            exhaust=self.server_pool_exhaust
        )
        return server_pool

    def create_ldap_server_obj(self, host: str) -> ldap3.Server:
        """
        Create ldap3 Server Object
        """
        server = ldap3.Server(
            host,
            port=self.server_port,
            use_ssl=self.server_use_ssl,
            connect_timeout=self.server_connect_timeout
        )
        return server

    def ldap_connection(self,
                        server_pool: ldap3.ServerPool,
                        username: str,
                        password: str) -> ldap3.Connection:
        """
        Create ldap(s) Connection Object
        """
        # attempt connection
        try:
            conn = ldap3.Connection(
                server_pool,
                user=username,
                password=password,
                auto_bind=ldap3.AUTO_BIND_NO_TLS,
                read_only=True,
                receive_timeout=self.server_receive_timeout)
        except ldap3.core.exceptions.LDAPBindError as exc:
            msg = "\n{exc_type}: {exc_msg}".format(
                exc_type=exc.__class__.__name__,
                exc_msg=exc.args[0] if exc.args else '')
            self.log.error("Failed to connect to ldap: {}".format(msg))
            conn = None
        return conn

    def get_nested_groups(self, conn, group: str) -> typing.List[str]:
        """
        Recursively search group for nested memberships
        """
        nested_groups = list()
        conn.search(
            search_base=self.group_search_base,
            search_filter=self.group_search_filter.format(group=group),
            search_scope=ldap3.SUBTREE)
        if conn.response:
            for nested_group in conn.response:
                if 'dn' in nested_group:
                    nested_groups.extend([nested_group['dn']])
                    groups = self.get_nested_groups(conn, nested_group['dn'])
                    nested_groups.extend(groups)
        nested_groups = list(set(nested_groups))
        return nested_groups

    def test_auth(self, conn: ldap3.Connection, auth_user_dn: str, password: str) -> bool:
        """
        Test User Authentication
        rebind ldap connection with authenticating user,
        gather results, and close connection
        """
        try:
            auth_bound = conn.rebind(user=auth_user_dn, password=password)
        except ldap3.core.exceptions.LDAPBindError:
            auth_bound = False
        finally:
            conn.unbind()
        return auth_bound

    @gen.coroutine
    def authenticate(self, handler: web.RequestHandler, data: dict) -> typing.Optional[str]:

        # define vars
        username = data['username']
        password = data['password']
        server_pool = self.create_ldap_server_pool_obj()
        conn_servers = list()

        # validate credentials
        username = username.lower()
        if not self.validate_username(username):
            self.log.error('Unsupported username supplied')
            return None
        if not password or not password.strip():
            self.log.error('Empty password supplied')
            return None

        # cast server_hosts to list
        if isinstance(self.server_hosts, str):
            self.server_hosts = self.server_hosts.split(',')

        # validate hosts and populate server_pool object
        for host in self.server_hosts:
            host = host.strip().lower()
            if not self.validate_host(host):
                self.log.warning(("Host '{}' not supplied in approved format. " +
                                  "Removing host from Server Pool").format(host))
                break
            server = self.create_ldap_server_obj(host)
            server_pool.add(server)
            conn_servers.extend([host])

        # verify ldap connection object parameters are defined
        if not server_pool.servers:
            self.log.error(
                "No hosts provided. ldap connection requires at least 1 host to connect to.")
            return None
        if self.bind_user_dn is None or not self.bind_user_dn.strip():
            self.log.error(
                "'bind_user_dn' config value undefined. required for ldap connection")
            return None
        if self.bind_user_password is None or not self.bind_user_password.strip():
            self.log.error(
                "'bind_user_password' config value undefined. required for ldap connection")
            return None

        # verify ldap search object parameters are defined
        if not self.user_search_base or not self.user_search_base.strip():
            self.log.error("'user_search_base' config value undefined. required for ldap search")
            return None
        if not self.user_search_filter or not self.user_search_filter.strip():
            self.log.error("'user_search_filter' config value undefined. required for ldap search")
            return None

        # open ldap connection and authenticate
        self.log.debug("Attempting ldap connection to {} with user '{}'".format(
            conn_servers, self.bind_user_dn))
        conn = self.ldap_connection(
            server_pool,
            self.bind_user_dn,
            self.bind_user_password)

        # proceed if connection has been established
        if not conn or not conn.bind():
            self.log.error(("Could not establish ldap connection to {} using '{}' " +
                            "and supplied bind_user_password.").format(
                                conn_servers, self.bind_user_dn))
            return None
        else:
            self.log.debug(
                "Successfully established connection to {} with user '{}'".format(
                    conn_servers, self.bind_user_dn))

            # format user search filter
            auth_user_search_filter = self.user_search_filter.format(
                username=username)

            # search for authenticating user in ldap
            self.log.debug("Attempting LDAP search using search_filter '{}'.".format(
                auth_user_search_filter))
            conn.search(
                search_base=self.user_search_base,
                search_filter=auth_user_search_filter,
                search_scope=ldap3.SUBTREE,
                attributes=self.user_membership_attribute if self.allowed_groups else list(),
                paged_size=2)

            # handle abnormal search results
            if not conn.response or len(conn.response) > 1:
                self.log.error(("LDAP search '{}' returned {} results. " +
                                "Please narrow search to 1 result").format(
                                    auth_user_search_filter, len(conn.response)))
                return None
            elif self.allowed_groups and 'attributes' not in conn.response[0].keys():
                self.log.error(("LDAP search '{}' did not return results for requested " +
                                "search attribute(s) '{}'").format(
                                    auth_user_search_filter, self.user_membership_attribute))
                return None
            else:
                self.log.debug("LDAP search '{}' found {} result(s).".format(
                    auth_user_search_filter, len(conn.response)))

                # copy response to var
                search_response = conn.response[0]

                # get authenticating user's ldap attributes
                if 'dn' not in search_response or not search_response['dn'].strip():
                    self.log.error(("Search results for user '{}' returned 'dn' attribute with " +
                                    "undefined or null value.").format(username))
                    conn.unbind()
                    return None
                else:
                    self.log.debug(
                        "Search results for user '{}' returned 'dn' attribute as '{}'".format(
                            username, search_response['dn']))
                    auth_user_dn = search_response['dn']

                # is authenticating user allowed
                if self.allowed_groups:
                    # compile list of user groups
                    if not search_response['attributes'][self.user_membership_attribute]:
                        self.log.error(("Search results for user '{}' returned '{}' attribute " +
                                        "with undefined or null value.").format(
                                            username, self.user_membership_attribute))
                        conn.unbind()
                        return None
                    else:
                        self.log.debug(
                            "Search results for user '{}' returned '{}' attribute as {}".format(
                                username, self.user_membership_attribute,
                                search_response['attributes'][self.user_membership_attribute]))
                        user_groups = search_response['attributes'][self.user_membership_attribute]

                    # compile list of permitted groups
                    permitted_groups = list()
                    permitted_groups.extend(self.allowed_groups)
                    if self.allow_nested_groups:
                        for group in self.allowed_groups:
                            nested_groups = self.get_nested_groups(conn, group)
                            permitted_groups.extend(nested_groups)

                    # is authenticating user a member of permitted_groups
                    allowed_memberships = list(set(user_groups).intersection(permitted_groups))
                    if allowed_memberships:
                        self.log.debug(("User '{}' found in the following allowed ldap groups " +
                                        "{}. Proceeding with authentication.").format(
                                            username, allowed_memberships))
                    else:
                        self.log.error(
                            "User '{}' is not a member of any permitted groups {}".format(
                                username, permitted_groups))
                        return None
                else:
                    self.log.debug(("User '{}' will not be verified against allowed_groups due " +
                                    "to feature short-circuiting. Proceeding with " +
                                    "authentication.").format(username))

                # return auth results
                auth_bound = self.test_auth(conn, auth_user_dn, password) or False
                if auth_bound:
                    self.log.info(
                        "User '{}' successfully authenticated against ldap server {}.".format(
                            username, conn_servers))
                    auth_response = username
                else:
                    self.log.error(
                        "User '{}' authentication failed against ldap server {}.".format(
                            conn_servers, auth_user_dn))
                    auth_response = None
                return auth_response
